### ISSUE TITLE
Suggestion: Bidirectional sync?

### ISSUE THREAD

**Issue by AlJoelson** [GedasFX/decky-cloud-save#2]
This is probably a really silly suggestion as I'm sure it was one of the first thing you'd considered, but is there any hope for this plugin to support bidirectional synchronisation? It'd be absolutely perfect then!

**Comment by GedasFX** [comment]
Hey, yeah this is something that I would maybe like to solve too, however, it adds complications that I was not able to find elegant solutions for in a timely manner. Maybe once I have more time or I find better ways, I will add this, but for v1, I would only like to make it only as a backup tool (good timing too, as my wine folder decided to delete itself and the only thing that kept my save files alive was this plugin). 

For reference here are the issues that need resolving:

1. Path synchronization. It is not guaranteed that both devices will have the same path for the same game file. Making two folders on different computers point to the same location is rather clunky from both technical and UI/UX perspective. 
  > One idea that has been floating around my head about how to do it, would be to "select" a path and give it a destination name. So for example, on steam deck you would select a local path and a remote path, and on the other device you would do the same or equivalent. Copying of files from there would be covered by a custom tool or the plugin (if implemented).
2. Sync timing. When would this sync occur? Manually? Before game start? I am not sure if I could cancel a game start with the plugin, however I could make a notification just after launching if data locally is not up to sync. Auto sync would just be too dangerous here as some games launch very fast.
3. Deleted files. Some games I found use more than one save file. So far I was not able to see that they delete them, but that may always be a possibility. The way this plugin is designed currently, is if you delete files, they are not deleted from remote. This is useful when you uninstall the game (or game deletes itself like it happened to me), but keep the sync path as then at least the files wouldn't get deleted. This is purely speculation at this point as I have no evidence to back up this concern, but it is a concern nonetheless.
4. Conflicting files. Honestly, this is the biggest one. The user must always be informed that files changed, as, if they are not careful (and assuming we have solved sync timing from point 2), they can lose progress.
  > Realistically, the only hard part here would be to check for updated files, and on game start look up if the registered paths are up to date. It would make more sense to add profiles to have paths bound to games themselves (I should do that as it currently checks all paths, not just the closed game). 

Well as I was writing this, I did get an idea, although not sure how quickly I could implement it. It would go like this:
1. Have users create profiles for games. I *should* be able to get currently selected game from library, and for sure I can get the currently running game (would need to figure out edge cases in case multiple games are running).
    * This profile should have bound "remote" path and "local" path. Remote path would be optional, if left unspecified it would just use the local path for it as well (how it works currently).
2. On game start run file checker for the game (to see if remote files have updated). If so, give user a warning that on next sync your files will get overwritten (would need to think of something for panic close prevention as no matter what, these files would be newer, feeling that this part can balloon in complexity really quickly).
3. Add ability to sync individual or global profiles manually to fetch data from remote. No automation for now.

With this approach it should be possible to do this, however, I really cannot give you an ETA.

**Comment by AlJoelson** [comment]
That's all good, like you said v1 does the trick simply ensuring that data isn't lost if something happens to the device!

**Comment by dJani97** [comment]
Being aware of the problems that bidirectional sync can cause, I enabled [rclone bisync](https://rclone.org/bisync/) by updating my Decky Cloud Save installation.

To do this, I replaced `"copy"` with `"bisync"` in `/home/deck/homebrew/plugins/decky-cloud-save/main.py`. The updated code looks like this:

```
    async def sync_now(self):
        logger.debug("Executing: sync_now()")
        self.current_sync = await asyncio.subprocess.create_subprocess_exec(rclone_bin, *["bisync", "--filter-from", cfg_syncpath_filter_file, "/", "backend:decky-cloud-save", "--copy-links"])
```

So far it seems to work well, syncing in both directions when I press the "Sync Now" button.

Note that according to rclone, this is an experimental feature:
```
NOTICE: bisync is EXPERIMENTAL. Don't use in production!
```

**Comment by GedasFX** [comment]
@dJani97 hey, was on vacation, did not see your post. I could probably add an 'experimental (use at your own risk)' feature which would instead of performing safe backups do bidirectional sync instead with your approach. It is experimental and I really don't trust non stable features, but if the people want it and it works, I don't see why not just add that into the plugin.

If we do this, the issues compound to this:

1. Path synchronization, would be the same. If you have multiple steam decks it would work fine, if you have something else, you could symlink just fine (testing needed).
2. Sync timing. Would add a toggle to disable automatic syncing on game exit and it could be done manually only. This would prevent accidental save overwrites. Otherwise its up to the user to click sync before starting the game. I think I should be able to check if remote has a newer version, and do this check on game start. Then user would have a reminder to sync or close the game. Really pick your poison thing. I do see that rclone would make patch files of the conflict and could be confusing to new users why their save was deleted (even though it was simply renamed and gave doesnt recognize that file anymore). At least it would override correctly.
3. Deleted files. Bisync would delete files on remote if they are deleted locally, which may be fine while its an 'experimental' feature. Not a big fan of this as I experience wine liking to just randomly reinstall itself for me which deletes files, however cloud providers seem to allow recovery for deleted files which I guess is fine.
4. Conflicting files. See [2], conflicting files get renamed and user would need to manually resolve the conflicts.

**Comment by EnsoBrush** [comment]
@GedasFX

1. Path synchronization - The way I see it, in addition to specifying the path to the save files, you also specify the game for which the synchronization will be done. There should be a simple way to list the games currently installed on Steam Deck, and from such a list you could choose which game the path applies to. And it would be even better if selecting the file path was from the menu assigned to the game. As it currently works in the steamgriddb plugin. Then you know what is synced and what game it applies to. 
2. Sync timing - What if you did a save file sync right after Steam Deck wakes up? Would it be possible? If not, manually pressing the sync button also sounds like something that will solve the problem. Pressing a single button before starting the game was exactly what I did when I thought this functionality was already built-in.
3. Deleted files - There is a plug-in that is called Steamback. I know that it's a completely different mechanism, but with this plug-in it is possible to restore an older save file. And this means that they have a backup of several versions of the file. Maybe it would be a solution to run on the latest save files by default, but to be able to keep, for example, several versions back. Or that it would also be possible to manually say that this one particular version of the save files should be kept separately, outside the synchronization system. 
4. Conflicting files - If the above solutions could be implemented successfully, I suspect that this might no longer be a problem.

**Comment by EnsoBrush** [comment]
@GedasFX 

2. If it were possible to implement what I described in point 1 in such a way that the save file is assigned to a specific game, then maybe it would be possible to detect when one accesses a game card on Steam Deck? And then check where the latest file is and possibly perform a synchronization. Also, in general, I think it would then be possible to inject sync status information into this view. Similar to how it currently works in the HLTB plugin.

**Comment by GedasFX** [comment]
I really would be overwhelmed to do all of this right now, the scope of this task is ballooning way faster than I can even comprehend, so we could first establish things that are absolutely necessary.

1. Bisync support. The whole idea is to support the rclone bisync mode. According to the docs this comes with limitations:
   1. First run and conflicts require the use of `--resync`. This effectively acts as 'if there are conflicts, use files in either local, or remote and overwrite everything else. This could be achieved by detecting if remote exists for first run, or have a toggle for user to 'force overwrite'. The user should then need to understand that this toggle is dangerous.
   2. If conflicts do exist it seems like we can then use steam cloud issue resolver like of 'use local' or 'use remote'. Merges are too confusing for the average user. Would need to make a custom UI for that too.
2. Sync check. It would likely just be `bisync --dry-run` and then parse the results. If a conflict is detected, use conflict resolution UI and cancel all sync, if no conflict (and sync is pending) carry on. This could also be used to notify that a sync is pending.

Personally I'd like to focus on these 2 features, as the rest are just fluff and can be done at a later date.

**Comment by GedasFX** [comment]
2.1 Oh I see what you mean. But this would require profiles either way and would like tho shelf it for now.
3 This would require versioning to be done on the user's side, or some versioning mechanism be in place by plugin. Versioning is also quite tricky as you need to make sure you don't use a lot of storage. Either way I'd like to leave versioning out of this scope.
4 Versioning does not solve conflicts, just look at git 😄, there are merges and conflicts all the time.

**Comment by EnsoBrush** [comment]
For my purposes, it would be enough for me to manually say “now synchronize the save files both ways” and I can handle it. The rest of my ideas are, as you yourself described, “just fluff and can be done at a later date”. :)

**Comment by wogglenet** [comment]
I spent a couple hours trying to figure out why my PC saves weren't syncing to the deck with this plugin. Sync usually means bidirectional! Hopefully you have time to do this some day, the plugin is great otherwise. I tried adding bisync as above but it gave me a sync failed error

**Comment by GedasFX** [comment]
Sorry guys, I think this deserves an update. I got into lots of draining stuff irl recently, that I just physically never had the time to focus up, sit down, do the changes, debug. and test.

It is passing over and I see light at end of the tunnel, but I don't think it will be for another 2 weeks at least. I haven't forgotten about this!

**Comment by jdawgzim** [comment]
What about using something like Syncthing?  Maybe that'd be a separate plug-in

**Comment by NL-TCH** [comment]
is there a branch for this? i want to help if possible :)

**Comment by GedasFX** [comment]
Hey thanks for asking @NL-TCH! Unfortunately I wasn't able to find time or motivation to make a proper go at this, however I did try to prototype something.

My idea right now is to make a separate plugin entirely and use some other tool that specializes in game synchronization, and just make steam deck wrapper on top of it. This would come with an advantage of allowing users to go to desktop mode for setup, which is good, as I found UI quite finicky on decky gaming mode (mostly due to my lack of skills to be fair).

Currently my eyes are on https://github.com/mtkennerly/ludusavi as it looks incredibly promising, last I tried i got rudimentary plugin working that just synced on startup but I never got to test cross device sync as didn't get windows version to work. Then I got disappointed, said I will try again bit later, and two months later here we are 😆.

**Comment by GedasFX** [comment]
Ultimately I do want to finish this, but I also do not want to invent my own sync systems, nor I want to undermine the original idea of the plugin.

If you guys want to help, we could try pick a manager that would work and go from there? Like I said before, ludasavi is kind of the best option for integration as it has a CLI, but there may be other options. Syncthing was also suggested, but I never got it to work as stable as I wanted to, and, well, if I cant figure out how a tool works, can't expect others to understand my explanations either 😅 

I will wrap up with my tasks soon I think and I will pick this up again. I want this done. Hopefully for real this time 😅

**Comment by GedasFX** [comment]
I could but I do not really want to turn a small passion project into a management hell XD At least I get paid for it at work lol.

Jokes aside I will try to find time until the weekend, and after that, I know I will have a bunch of free time starting September. I opened up a new issue https://github.com/GedasFX/decky-cloud-save/issues/37 where we could discuss that.

**Comment by jdawgzim** [comment]
Nice guide for setting up Syncthing as service to sync folders:
https://youtu.be/v_Hh1YDOf8s 

Then use https://www.pcgamingwiki.com/wiki/Home to find the game folder you want to sync.

**Comment by dyllan500** [comment]
I am not sure what you guys have in mind on how you want this feature to work.

I was testing out bisync this is what I found. If I sync on my Steam Deck it works on updating the game save on the server. But if I then sync on my Living room PC it will overwrite the game save on the server with an older version instead of updating the local game save  on the PC.

 I personally would like it if all things where in sync across multiple devices, but that will defiantly add a lot more complexity to this project.
 
 I thinking of forking and tweaking the plugin to update the server copy after exiting any game. The sync now button will only every grab the remote copy of the files and update the local ones.

**Comment by AkazaRenn** [comment]
I'm wondering if we can add a button to manually trigger a download or bidirectional sync, would be exceptionally helpful when setting up a new machine.

**Comment by t3nk3y** [comment]
We've read through this feature request and the PR code that NL-TCH worked on.  If you are running in to issues, we think the real problem is that the sync's need to happen both pre-launch, and post-exit.  As is, you would need to manually trigger the sync before launching, or a cloud save will not get synced to the deck(or other systems) you will then play, it will generate game data, and when you exit the game, it could overwrite what's in the cloud(even if you exit right away after realizing your mistake, any file that had it's timestamp updated will overwrite).

Does anyone know if there is a hook in Steam that could be used to run this pre-launch, as well?

**Comment by GedasFX** [comment]
There actually is a pre-launch hook, but it doesn't block the launch itself to my knowledge so its usability is limited.

Actually the main issue is a bit more bureaucratic right now. The issue is that to support bisync you need to run a bunch of custom commands to get debug output, get merge conflict resolutions, etc. - things that I do not think I am capable of showing to the user in a friendly manner.

Ultimately I would like to publish the plugin to the store, and for that, I would need this to somehow pass testing, and its hard to test hidden non-user friendly features. I definitely would not be able to present this to tester team and in good conscious say "yeah this is ready, please verify".

So this feature is basically stuck. I do not know how to properly display the intricacies of bisync error handling, and I do not know how to properly convey to testers how to verify the feature actually works and is release worthy. Maybe a good documentation may help, but even when testing locally it felt like there were unlimited cases to be supported, which simply overwhelmed me.

I really hope I could push the other plugin out and solve this issue that way, but for real life reasons, this has proven to be difficult.

**Comment by t3nk3y** [comment]
I was looking through some other plugin code and realized that, from the
plugin that pauses games on opening, I realized we could detect when the
game is being started, "pause" the game, do sync, then unpause.

That said, we also thought through what the process should really look like
there, and we think it makes more sense to pull before launch and push
after, and not use the bisync method at all.  Should also have a versions
folder, prerun the pull with preview only, move any files that are to be
replaced to versions, then do the real pull, and same sort of thing should
happen on push.

Like you, I've too much going on in our life to work on this right now, so
as much as this would be awesome to work on, and for folks to have, with
syncthing doing the sync work and already doing versions and
bidirectionality and conflict resolution, and since we also have the local
folders set to sync up to our cloud which gets backed up as well, we are
personally covered well for our use case

Would be open to helping with problem solving, code and logic suggestions
and testing, but don't have the time to work on this directly.

**Comment by GedasFX** [comment]
https://github.com/SteamDeckHomebrew/decky-plugin-database/pull/637
ITS DONE. BYE. CLOSED. MOVING ON!
